### ISSUE TITLE
Silent truncating large identifiers

### ISSUE THREAD

**Issue by otaviojacobi** [elixir-ecto/ecto#4388]
### Elixir version

Elixir 1.16.1 (compiled with Erlang/OTP 26)

### Database and Version

PostgreSQL 16

### Ecto Versions

3.11.1

### Database Adapter and Versions (postgrex, myxql, etc)

postgrex 0.17.4

### Current behavior

By default, PostgreSQL truncates identitfier (table names, fkey names etc) on 63 bytes, from their docs

“The system uses no more than NAMEDATALEN -1 bytes of an identifier; longer names can be written in commands, but they will be truncated. By default, NAMEDATALEN is 64 so the maximum identifier length is 63 bytes.”

Similar things happen on other DBs https://dev.mysql.com/doc/refman/8.3/en/identifier-length.html

And this can generate some undesired behaviour when using Ecto changesets default names. For example, here: [ecto/lib/ecto/changeset.ex at master · elixir-ecto/ecto · GitHub](https://github.com/elixir-ecto/ecto/blob/master/lib/ecto/changeset.ex#L3792)

If you don’t manually choose a name for your FKEY, Ecto will keep thing it is called “my_very_very_very_very_long_and_very_explicit_and_even_longer_fkey” but postgres will silently save it as “my_very_very_very_very_long_and_very_explicit_and_even_longer_f” and thus ecto gets lost.

Workarounds exist (e.g. explicitly naming the constraint or increasing the max identifier on a DB level), however, I would have expected Ecto to error instead of just silently creating a fkey with a different name and then trying to user wrong name over and over.

### Expected behavior

Fail the migration/constraint creation if the default FKEY name assigned by Ecto does not match the actual value on the DB.

I am happy to PR a fix if we can agree on which approach should Ecto follow for this kind of errors.

See: https://elixirforum.com/t/dealing-with-truncating-identifiers-on-db/62052

**Comment by josevalim** [comment]
Doesn't the database log when it truncates? I thought we were reporting these logs. :(

**Comment by otaviojacobi** [comment]
Running the following migration on a almost empty repo (just after setting up an empty project):
```
❯ mix ecto.gen.migration create_org
```
```elixir
defmodule Friends.Repo.Migrations.CreateOrg do
  use Ecto.Migration

  def change do
    create table(:organizations) do
      add :subscription_billing_provider_organization_id, references(:organizations, on_delete: :nilify_all)
    end
  end
end
```

I get:

```
* creating priv/repo/migrations/20240303172204_create_org.exs
❯ mix ecto.migrate

14:23:08.647 [info] == Running 20240303172204 Friends.Repo.Migrations.CreateOrg.change/0 forward

14:23:08.648 [info] create table organizations

14:23:08.654 [info] == Migrated 20240303172204 in 0.0s
```

So I don't see any logs (but maybe I am missing enabling debug logs somewhere?). This example creates a FK constraint named "organizations_subscription_billing_provider_organization_id_fke"

If I manually run directly on postgres DB: 
```
ALTER TABLE "organizations"
ADD CONSTRAINT "my_very_very_very_very_long_and_very_explicit_and_even_longer_fkey" CHECK (id < 5);
```

I do see a message on postgresql logs: `NOTICE:  identifier "my_very_very_very_very_long_and_very_explicit_and_even_longer_fkey" will be truncated to "my_very_very_very_very_long_and_very_explicit_and_even_longer_f"`

**Comment by josevalim** [comment]
Yeah, I would investigate then why we are not getting those notices. :+1:

**Comment by otaviojacobi** [comment]
Just reporting in my findings so far... This seems to be coming from `postgrex`
1. postgrex correctly sets the [status.messages](https://github.com/elixir-ecto/postgrex/blob/e12da3451336032341e04da1759ad25b6a491ccf/lib/postgrex/protocol.ex#L3318) here, insepcting the `status.message` yields: 
```elixir
{:ok, {:msg_notice, %{code: "42622", line: "99", message: "identifier \"organizations_subscription_billing_provider_organization_id_fkey\" will be truncated to \"organizations_subscription_billing_provider_organization_id_fke\"", file: "scansup.c", unknown: "NOTICE", severity: "NOTICE", routine: "truncate_identifier"}}, ""}
```

As expected. However, this is happening `before` execution, e.g. on the `db_connection` https://github.com/elixir-ecto/db_connection/blob/8ef1f2ea54922873590b8939f2dad6b031c5b49c/lib/db_connection.ex#L1510-L1527

My feeling so far is that somehow the `status.messages` populated on the `run_prepare` are not being forwarded to the `run_ execute` and therefore never reaching the final destination. I might have some time tonight to continue investigating.

**Comment by otaviojacobi** [comment]
@josevalim Yep - I confirmed behaviour above. Basically, none of the `prepare` logs are being forwarded. My understanding is that the `messages` sent by postgresql, even in the "prepare" phase, should be bubbled up, however, the `%Query{}` does not contain a "messages" only the result. My understanding is that somehow the messages on the created status for the prepare (or the entire `status`) should go into the execute phase. Any ideas on how I could move this forward keeping the current interface?

**Comment by josevalim** [comment]
@otaviojacobi great find. Now it was trivial to add a fix: fcad612213ac5f793b4ec41696b6ee8320ea9315 :)
